spec/unit/ace/file_mutex_spec.rb

```diff
@@ -0,0 +1,188 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+require 'ace/file_mutex'
+require 'timeout'
+
+RSpec.describe ACE::FileMutex do
+  let(:lock_file) { 'spec/fixtures/test.lock' }
+  let(:mutex) { described_class.new(lock_file) }
+  let(:lock_content) { [] }
+
+  before do
+    File.delete(lock_file) if File.exist?(lock_file)
+  end
+
+  describe '#with_read_lock' do
+    let(:p1_message) { 'process 1 written' }
+    let(:p2_message) { 'process 2 written' }
+
+    it 'creates the lock file' do
+      foo = mutex.with_read_lock do
+        lock_content
+      end
+
+      expect(File).to exist(lock_file)
+      expect(foo).to eq(lock_content)
+    end
+
+    it 'allows multiple non-blocking calls' do
+      p1_in_read, _p1_in_write = IO.pipe
+      p1_out_read, p1_out_write = IO.pipe
+      p2_in_read, _p2_in_write = IO.pipe
+      p2_out_read, p2_out_write = IO.pipe
+
+      Timeout.timeout(2) do
+        fork do
+          mutex.with_read_lock do
+            p1_out_write.puts p1_message
+
+            # start the blocking read
+            puts p1_in_read.read
+
+            # tidy up
+            p1_out_write.close
+          end
+        end
+        fork do
+          mutex.with_read_lock do
+            p2_out_write.puts p2_message
+
+            # start the blocking read
+            puts p2_in_read.read
+
+            # tidy up
+            p2_out_write.close
+          end
+        end
+        # close the blocking reads
+        p1_in_read.close
+        p2_in_read.close
+
+        # now check that each fork performed the operations at the same time
+        expect(p1_out_read.gets).to match(/#{p1_message}/)
+        expect(p2_out_read.gets).to match(/#{p2_message}/)
+      end
+
+      # tidy up
+      p1_out_read.close
+      p2_out_read.close
+    end
+  end
+
+  describe '#with_write_lock' do
+    it 'creates the lock file' do
+      foo = mutex.with_write_lock do
+        lock_content
+      end
+
+      expect(File).to exist(lock_file)
+      expect(foo).to eq(lock_content)
+    end
+
+    context 'when a write lock is held' do
+      let(:p1_message) { 'process 1 written' }
+      let(:p2_message) { 'process 2 should not write' }
+
+      p1_in_read, _p1_in_write = IO.pipe
+      p1_out_read, p1_out_write = IO.pipe
+      p2_in_read, _p2_in_write = IO.pipe
+      p2_out_read, p2_out_write = IO.pipe
+      exception_thrown = false
+
+      it 'second lock is blocked' do
```

Comment:
this test message is better, but can still be improved. e.g. "shared lock blocks when exclusive lock is taken"